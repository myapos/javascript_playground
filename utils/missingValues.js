import verboseLog from './verboseLog';
import groupByProperty from './groupByProperty';
import { first } from 'lodash';

/**Detect a leap year */
function leapYear(year) {
  return (year % 4 == 0 && year % 100 != 0) || year % 400 == 0;
}
/* It adds a zero if value is smaller than ten. This will help to keep unique values in Map */
const prettyValue = (value) => {
  let pretty = parseFloat(value);
  if (pretty < 10) {
    pretty = `0${pretty}`;
  }
  return pretty;
};
const monthInfo = {
  '01': {
    month: 'January',
    numOfDays: 31,
    order: '01',
  },
  '02': {
    month: 'February',
    numOfDays: 28,
    order: '02',
  },
  '03': {
    month: 'March',
    numOfDays: 31,
    order: '03',
  },
  '04': {
    month: 'April',
    numOfDays: 30,
    order: '04',
  },
  '05': {
    month: 'May',
    numOfDays: 31,
    order: '05',
  },
  '06': {
    month: 'June',
    numOfDays: 30,
    order: '06',
  },
  '07': {
    month: 'July',
    numOfDays: 31,
    order: '07',
  },
  '08': {
    month: 'August',
    numOfDays: 31,
    order: '08',
  },
  '09': {
    month: 'September',
    numOfDays: 30,
    order: '09',
  },
  '10': {
    month: 'October',
    numOfDays: 31,
    order: '10',
  },
  '11': {
    month: 'November',
    numOfDays: 30,
    order: '11',
  },
  '12': {
    month: 'December',
    numOfDays: 31,
    order: '12',
  },
};

/** It accepts a day and tests if the next date exist in the initial data */
const nextDateExists = (raw, dayOfMonth, monthIs, dates, year) => {
  //   console.log('---------', dayOfMonth, monthIs, dates);
  let nextDay = ++dayOfMonth;

  if (nextDay >= monthIs.numOfDays) {
    nextDay = monthIs.numOfDays;
  }

  const nextDate = `${prettyValue(nextDay)}-${monthIs.order}-${year}`;
  return {
    exists: dates.includes(nextDate),
    nextDate,
  };
};

const splitDate = (date) => date.split('-');

const convertArrayToMap = (ar) => {
  let myMap = new Map();

  ar.forEach((raw) => {
    const splitted = splitDate(raw.Date);

    // detect key
    const prettyKey = prettyValue(splitted[1]);

    myMap.set(raw.Date, {
      ...raw,
      filled: typeof raw.filled !== 'undefined' ? raw.filled : false,
      month: monthInfo[prettyKey].month,
      order: monthInfo[prettyKey].order,
    });
  });

  return myMap;
};

const convertMapToArray = (myMap) => {
  const ar = [];
  for (let [key, value] of myMap) {
    // console.log(key + ' = ' + JSON.stringify(value));
    ar.push({
      Date: key,
      Kgrs: value.Kgrs,
      month: value.month,
      order: value.order,
      filled: value.filled,
    });
  }
  // console.log('converted array:', JSON.stringify(ar, null, 2));
  // console.log('converted array:', ar);

  return ar;
};

/** It detects missing values in two modes (forward and reverse). In forward mode it scans the dates inside a
 * year and it fills in the missing dates with the value of previous day. In reverse mode it scans the  data in
 * reverse and fills the missing date with the next date. The result should be a range with full dates containing
 * values from prexisted data
 *
 * rawData: raw data from parsing csv. it is an array of json objects
 * dates: the array of dates extracted from rawData
 * mode: can be 'forward' or 'reverse' string
 */
const missingValues = (rawData, dates, mode) => {
  let converted = [];
  let grouped;
  verboseLog('---------detectMissingValues-------------- ' + mode);
  const filledMap = convertArrayToMap(rawData);

  // get the first month in the values
  let tempMonth = rawData[0].Date.split('-')[1];
  let tempMonthIs = monthInfo[tempMonth];

  if (mode === 'forward') {
    // loop
    rawData.forEach((raw, index) => {
      // detect dayOfMonth
      const dayOfMonth = parseFloat(raw.Date.split('-')[0]);

      // check for the first day in each month
      // detect month

      const month = raw.Date.split('-')[1];
      const monthIs = monthInfo[month];

      // detect year
      const year = raw.Date.split('-')[2];

      // build nextDate and search if it exists in initial dates
      let { exists, nextDate } = nextDateExists(raw.Date, dayOfMonth, monthIs, dates, year);

      let monthOfNextDate = nextDate.split('-')[1];
      let nextDayIsInSameMonth = monthOfNextDate === monthIs.order;
      let counter = 0;

      // if the next date does not exist in the array then add it
      while (!exists && nextDayIsInSameMonth && counter <= monthIs.numOfDays) {
        // push missing date with the previous value

        const splittedKey = nextDate.split('-');

        const prettyDate = `${prettyValue(splittedKey[0])}-${prettyValue(
          splittedKey[1],
        )}-${prettyValue(splittedKey[2])}`;

        filledMap.set(prettyDate, {
          Date: prettyDate,
          Kgrs: raw.Kgrs,
          filled: true,
          month: monthInfo[prettyValue(splittedKey[1])].month,
          order: monthInfo[prettyValue(splittedKey[1])].order,
        });
        monthOfNextDate = nextDate.split('-')[1];
        const nextDayOfMonth = prettyValue(parseFloat(nextDate.split('-')[0]));
        nextDayIsInSameMonth = monthOfNextDate === monthIs.order;
        ({ exists, nextDate } = nextDateExists(nextDate, nextDayOfMonth, monthIs, dates, year));

        counter++;
      }

      tempMonthIs = { ...monthIs };
    });

    // sort by date
    var mapAsc = new Map(
      [...filledMap.entries()].sort(function (a, b) {
        // Turn your strings into dates, and then subtract them
        // to get a value that is either negative, positive, or zero.
        const splittedA = splitDate(a[0]);
        const splittedB = splitDate(b[0]);
        const DateA = new Date(splittedA[1] + '-' + splittedA[0] + '-' + splittedA[2]);
        const DateB = new Date(splittedB[1] + '-' + splittedB[0] + '-' + splittedB[2]);
        // console.log('DateA', DateA);
        return DateA - DateB;
      }),
    );

    converted = convertMapToArray(mapAsc);
  } else if (mode === 'reverse') {
    const extraInfoArray = convertMapToArray(filledMap);

    // group by month
    grouped = groupByProperty(extraInfoArray, 'month');

    // console.log('grouped', grouped);
    // loop in reverse mode for each month and detect if there are missing values for each month
    // in the beginning
    let hasMissingValuesPerMonth = false;
    Object.keys(grouped).forEach((month) => {
      const splitted = splitDate(grouped[month][0].Date);
      const monthKey = splitted[1];

      if (grouped[month].length !== monthInfo[monthKey].numOfDays) {
        const monthData = grouped[month];
        hasMissingValuesPerMonth = true;
        const firstMonthRecording = monthData[0];

        const numOfMissingDays = monthInfo[monthKey].numOfDays - grouped[month].length;
        // console.log('----------monthData----------', monthData);
        // console.log('first recorded value', firstMonthRecording);

        console.log(`month:${month} has ${numOfMissingDays} missing days in the beginning`);

        const splitted = splitDate(firstMonthRecording.Date);
        const firstDay = parseFloat(splitted[0]);

        // loop and fill missing dates in the beginning
        for (let i = 1; i <= numOfMissingDays; i++) {
          const dayNumber = prettyValue(firstDay - i);
          // build date
          const newDate = `${dayNumber}-${splitted[1]}-${splitted[2]}`;
          grouped[month].unshift({
            ...firstMonthRecording,
            Date: newDate,
            filled: true,
          });
        }
      }
    });

    // console.log('-----grouped--------', JSON.stringify(grouped));
    // flaten grouped to converted

    Object.keys(grouped).forEach((month) => {
      converted.push(...grouped[month]);
    });
  }

  // console.log('grouped', grouped);
  return converted;
};

export default missingValues;
