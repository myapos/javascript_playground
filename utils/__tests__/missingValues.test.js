import missingValues from '../missingValues';
import { rawData } from './__data__/rawData';
import { dates } from './__data__/dates';

describe('missingValues', () => {
  it('should fill missing values with the previous existing value', () => {
    const filledValues = missingValues(rawData, dates, 'forward');
    expect(filledValues).toMatchSnapshot();
  });
});
