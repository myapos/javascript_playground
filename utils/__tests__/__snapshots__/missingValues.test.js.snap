// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`missingValues should fill missing values with the previous existing value 1`] = `
Array [
  Object {
    "Date": "14-03-2020",
    "Kgrs": "108",
    "filled": false,
    "month": "March",
    "order": "03",
  },
  Object {
    "Date": "15-03-2020",
    "Kgrs": "108",
    "filled": true,
    "month": "March",
    "order": "03",
  },
  Object {
    "Date": "16-03-2020",
    "Kgrs": "108",
    "filled": true,
    "month": "March",
    "order": "03",
  },
  Object {
    "Date": "17-03-2020",
    "Kgrs": "108",
    "filled": true,
    "month": "March",
    "order": "03",
  },
  Object {
    "Date": "18-03-2020",
    "Kgrs": "108",
    "filled": true,
    "month": "March",
    "order": "03",
  },
  Object {
    "Date": "19-03-2020",
    "Kgrs": "108",
    "filled": true,
    "month": "March",
    "order": "03",
  },
  Object {
    "Date": "20-03-2020",
    "Kgrs": "108",
    "filled": true,
    "month": "March",
    "order": "03",
  },
  Object {
    "Date": "21-03-2020",
    "Kgrs": "108",
    "filled": true,
    "month": "March",
    "order": "03",
  },
  Object {
    "Date": "22-03-2020",
    "Kgrs": "108",
    "filled": true,
    "month": "March",
    "order": "03",
  },
  Object {
    "Date": "23-03-2020",
    "Kgrs": "108",
    "filled": true,
    "month": "March",
    "order": "03",
  },
  Object {
    "Date": "24-03-2020",
    "Kgrs": "108",
    "filled": true,
    "month": "March",
    "order": "03",
  },
  Object {
    "Date": "25-03-2020",
    "Kgrs": "108",
    "filled": true,
    "month": "March",
    "order": "03",
  },
  Object {
    "Date": "26-03-2020",
    "Kgrs": "108",
    "filled": true,
    "month": "March",
    "order": "03",
  },
  Object {
    "Date": "27-03-2020",
    "Kgrs": "108",
    "filled": true,
    "month": "March",
    "order": "03",
  },
  Object {
    "Date": "28-03-2020",
    "Kgrs": "108",
    "filled": true,
    "month": "March",
    "order": "03",
  },
  Object {
    "Date": "29-03-2020",
    "Kgrs": "108",
    "filled": true,
    "month": "March",
    "order": "03",
  },
  Object {
    "Date": "30-03-2020",
    "Kgrs": "108",
    "filled": true,
    "month": "March",
    "order": "03",
  },
  Object {
    "Date": "31-03-2020",
    "Kgrs": "108",
    "filled": true,
    "month": "March",
    "order": "03",
  },
  Object {
    "Date": "03-05-2020",
    "Kgrs": "104,9",
    "filled": false,
    "month": "May",
    "order": "05",
  },
  Object {
    "Date": "04-05-2020",
    "Kgrs": "104,9",
    "filled": true,
    "month": "May",
    "order": "05",
  },
  Object {
    "Date": "05-05-2020",
    "Kgrs": "104,9",
    "filled": true,
    "month": "May",
    "order": "05",
  },
  Object {
    "Date": "06-05-2020",
    "Kgrs": "104,9",
    "filled": true,
    "month": "May",
    "order": "05",
  },
  Object {
    "Date": "07-05-2020",
    "Kgrs": "104,9",
    "filled": true,
    "month": "May",
    "order": "05",
  },
  Object {
    "Date": "08-05-2020",
    "Kgrs": "104,9",
    "filled": true,
    "month": "May",
    "order": "05",
  },
  Object {
    "Date": "09-05-2020",
    "Kgrs": "104,9",
    "filled": true,
    "month": "May",
    "order": "05",
  },
  Object {
    "Date": "10-05-2020",
    "Kgrs": "104,9",
    "filled": true,
    "month": "May",
    "order": "05",
  },
  Object {
    "Date": "11-05-2020",
    "Kgrs": "104,9",
    "filled": true,
    "month": "May",
    "order": "05",
  },
  Object {
    "Date": "12-05-2020",
    "Kgrs": "103,6",
    "filled": false,
    "month": "May",
    "order": "05",
  },
  Object {
    "Date": "13-05-2020",
    "Kgrs": "103,6",
    "filled": true,
    "month": "May",
    "order": "05",
  },
  Object {
    "Date": "14-05-2020",
    "Kgrs": "103,6",
    "filled": true,
    "month": "May",
    "order": "05",
  },
  Object {
    "Date": "15-05-2020",
    "Kgrs": "103,6",
    "filled": true,
    "month": "May",
    "order": "05",
  },
  Object {
    "Date": "16-05-2020",
    "Kgrs": "103,6",
    "filled": true,
    "month": "May",
    "order": "05",
  },
  Object {
    "Date": "17-05-2020",
    "Kgrs": "103,6",
    "filled": true,
    "month": "May",
    "order": "05",
  },
  Object {
    "Date": "18-05-2020",
    "Kgrs": "104,7",
    "filled": false,
    "month": "May",
    "order": "05",
  },
  Object {
    "Date": "19-05-2020",
    "Kgrs": "104,2",
    "filled": false,
    "month": "May",
    "order": "05",
  },
  Object {
    "Date": "20-05-2020",
    "Kgrs": "104,2",
    "filled": false,
    "month": "May",
    "order": "05",
  },
  Object {
    "Date": "21-05-2020",
    "Kgrs": "104,2",
    "filled": false,
    "month": "May",
    "order": "05",
  },
  Object {
    "Date": "22-05-2020",
    "Kgrs": "103,7",
    "filled": false,
    "month": "May",
    "order": "05",
  },
  Object {
    "Date": "23-05-2020",
    "Kgrs": "103,7",
    "filled": false,
    "month": "May",
    "order": "05",
  },
  Object {
    "Date": "24-05-2020",
    "Kgrs": "103,2",
    "filled": false,
    "month": "May",
    "order": "05",
  },
  Object {
    "Date": "25-05-2020",
    "Kgrs": "102,8",
    "filled": false,
    "month": "May",
    "order": "05",
  },
  Object {
    "Date": "26-05-2020",
    "Kgrs": "103,1",
    "filled": false,
    "month": "May",
    "order": "05",
  },
  Object {
    "Date": "27-05-2020",
    "Kgrs": "103,1",
    "filled": true,
    "month": "May",
    "order": "05",
  },
  Object {
    "Date": "28-05-2020",
    "Kgrs": "103,1",
    "filled": true,
    "month": "May",
    "order": "05",
  },
  Object {
    "Date": "29-05-2020",
    "Kgrs": "103,1",
    "filled": true,
    "month": "May",
    "order": "05",
  },
  Object {
    "Date": "30-05-2020",
    "Kgrs": "103,1",
    "filled": true,
    "month": "May",
    "order": "05",
  },
  Object {
    "Date": "31-05-2020",
    "Kgrs": "103,1",
    "filled": true,
    "month": "May",
    "order": "05",
  },
]
`;
